### DISCUSSION TITLE
Segmentation fault when performing very large number of operations

### DISCUSSION THREAD

**Discussion by RiskoZS** [libvips/libvips#3915] [General]
**Describe the bug**
When applying many (~50,000) operations to an image, libvips experiences a segfault.

**To Reproduce**
I can consistently reproduce the segfault with this Python script:
```py
from pyvips import Image

img = Image.black(500, 500)

for i in range(50000):
    img = img.sin()

print("Done")
img = img.copy_memory()
```

**Expected behavior**
The above program prints "Done" and eventually exits quietly or throws a Python error.

**Actual behavior**
The above program prints "Done" and then throws a segmentation fault:
```
Done
[1]    142277 segmentation fault (core dumped)  python3 seg.py
```

**Screenshots**
N/A

**Environment**
 - OS: Fedora 39
 - Vips: master (8.16.0), but can also reproduce it on 8.15.1 (haven't tested other versions)

**Additional context**
It's most likely not pyvips' fault since I can get the segfault to occur using the Lua binding too.

**Comment by jcupitt** [comment]
Hi @RiskoZoSlovenska,

This is expected behaviour. libvips isn't calculating `sin` each loop, instead it's adding another node to a graph attached to `img`. When you finally force evaluation with `copy_memory`, it walks the graph computing pixel regions, and that triggers a C stack overflow.

Try this:

```python
#!/usr/bin/env python3

import pyvips

img = pyvips.Image.black(500, 500)

for i in range(5000):
    print(f"loop {i} ...")
    for _ in range(1000):
        img = img.sin()

    img = img.copy_memory()
```

Now no pipeline gets longer than 1000 nodes and it'll run forever.

**Reply by RiskoZS** [reply]
Wouldn't it be better if libvips tried to catch such crashes though? Perhaps by limiting how long pipelines can get or by checking whether there's enough space before walking the graph, if that's even possible?
